### ISSUE TITLE
insert_pdf: "Too many open files" when looping

### ISSUE THREAD

**Issue by cbm755** [pymupdf/PyMuPDF#1417]
I'm experiencing something like #1351 and trying to break it down into a couple MWE.

This code looks making new documents and inserting a PDF file into them.  After about 1020, it dies
```
for n in range(0, 1500):
    print(n)
    doc = fitz.open()
    cover = fitz.open("cover.pdf")
    doc.insert_pdf(cover)
    doc.save(f"out{n:04}.pdf")
```
The error is:
```
1017
1018
1019
1020
mupdf: cannot open file 'out1020.pdf': Too many open files
Traceback (most recent call last):
  File "/home/cbm/.local/lib/python3.10/site-packages/fitz/fitz.py", line 4255, in save
RuntimeError: cannot open file 'out1020.pdf': Too many open files
```

Most likely my kernel limits open files to 1024 per process.  stdout, stdin, stderr, then rest exhausted.

I tested this with 1.18.17 and 1.19.2.

If we instead open `cover.pdf` directly instead of using `insert_pdf` we have no such error:
```
for n in range(0, 1500):
    print(n)
    doc = fitz.open("cover.pdf")
    doc.save(f"out{n:04}.pdf")
```

PyMuPDF 1.18.17, 1.19.2 both from `pip install`, so **wheel**.
```
3.10.0 (default, Oct  4 2021, 00:00:00) [GCC 11.2.1 20210728 (Red Hat 11.2.1-1)] 
 linux 
 
PyMuPDF 1.19.2: Python bindings for the MuPDF 1.19.0 library.
Version date: 2021-11-20 00:00:01.
Built for Python 3.10 on linux (64-bit).
```

**Comment by cbm755** [comment]
Oops, should've mentioned, adding explicit `doc.close()` doesn't help.

**Comment by cbm755** [comment]
Ah!  So I tried:
```
for n in range(0, 1500):
    print(n)
    doc = fitz.open()
    cover = fitz.open("cover.pdf")
    doc.insert_pdf(cover)
    cover.close()
    doc.save(f"out{n:04}.pdf")
    doc.close()
```
And this works fine.  I think this is PEBKAC error.  `fitz.open` is like Python open: you must explicitly close the result otherwise file handles stay in use.

**Comment by cbm755** [comment]
Actually I just re-read #533 now I'm confused again.  Back then I took out all our explicit `.close()` as being unnecessary, and causing crashes on a specific and old version of pymupdf.  We're pinned much higher than 1.17.1 now so its no problem to put `.close()` back but I'm confused about what is "right" here.

Also both of these context manager approaches lead to exhausted file handles:
```
for n in range(0, 1500):
    print(n)
    with fitz.open() as doc:
        with fitz.open("cover.pdf") as cover:
            doc.insert_pdf(cover)
            doc.save(f"out{n:04}.pdf")

for n in range(0, 1500):
    print(n)
    with fitz.open() as doc, fitz.open("cover.pdf") as cover:
        doc.insert_pdf(cover)
        doc.save(f"out{n:04}.pdf")
```
perhaps I should file a separate issue for that?

**Comment by cbm755** [comment]
btw, is there a test suite?  The non-exausted file-handles when using context manager seems like something I could write a unit test for, but checking that into Plom doesn't seem quite right.

**Comment by JorjMcKie** [comment]
> btw, is there a test suite? The non-exausted file-handles when using context manager seems like something I could write a unit test for, but checking that into Plom doesn't seem quite right.

I am using `pytest` exclusively. The tests are located in the `tests` folder of the repository. So you could add a script via a PR there.

**Comment by JorjMcKie** [comment]
New version 1.19.3 is being uploaded to PyPI.

**Comment by lidacity** [comment]
not fixed!

```
import fitz

Result = fitz.open()

for n in range(0, 1500):
    print(n)
    with fitz.open("test.pdf") as PDF:
        Result.insertPDF(PDF)
```

The error is (1.18.15 and after):
```
1019
1020
1021
mupdf: cannot open test.pdf: Too many open files
Traceback (most recent call last):
  File "test.py", line 8, in <module>
  File "/home/dzmitry/CM/env/lib/python3.8/site-packages/fitz/fitz.py", line 3671, in __init__
RuntimeError: cannot open test.pdf: Too many open files
```
No error in 1.18.14 (and before), and 1.19.3 (only!)
On 1.19.4 - The error is:
```
1019
1020
1021
mupdf: cannot open test.pdf: Too many open files
Traceback (most recent call last):
  File "test.py", line 8, in <module>
  File "/home/dzmitry/CM/env/lib/python3.8/site-packages/fitz/fitz.py", line 3845, in __init__
RuntimeError: cannot open test.pdf: Too many open files
```

**Comment by cbm755** [comment]
Sadly, can confirm, @JorjMcKie sorry for not doing that unit test I promised!

**Comment by JorjMcKie** [comment]
@cbm755 / @lidacity 

Found the problem, generation of pre-wheels for Linux and OSX are underway.
I'll let you know when that is finished.

Background:
In an effort to get rid of the previously multiple places where document objects are destroyed, I forgot to put document dropping on the C level in a ``while`` loop (keep dropping while reference count > 0).
Sorry about that.

@cbm755 - no worries, I will add a respective test script to the collection.

**Comment by JorjMcKie** [comment]
Please find the mentioned wheels [here](https://github.com/pymupdf/PyMuPDF-wheels/tree/master/wheelhouse).

**Comment by JorjMcKie** [comment]
Fixed in v1.19.5 currently being uploaded.
